srtcore/core.cpp

```diff
@@ -6827,85 +6861,17 @@ void CUDT::processCtrl(CPacket& ctrlpkt)
       * which may go crazy and stay there, preventing proper stream recovery.
       */
 
-      if (CSeqNo::seqoff(m_iSndLastFullAck, ack) <= 0)
+      if (CSeqNo::seqoff(m_iSndLastFullAck, ackdata_seqno) <= 0)
       {
          // discard it if it is a repeated ACK
          CGuard::leaveCS(m_AckLock);
          break;
       }
-      m_iSndLastFullAck = ack;
-
-      int offset = CSeqNo::seqoff(m_iSndLastDataAck, ack);
-      // IF distance between m_iSndLastDataAck and ack is nonempty...
-      if (offset > 0) {
-          // acknowledge the sending buffer (remove data that predate 'ack')
-          m_pSndBuffer->ackData(offset);
```

Comment:
It turned out thee was an issue. After the `m_pSndBuffer` acknowledges packets, several line below the epoll updates the state of the socket, signaling there is available space in the sender's buffer to write (submit) more data.
But the access to epoll is not protected, and at the same time the writing thread (`CUDT::sendmsg2`) may be already writing data to it.
In that case the following call, marking the socket as available for writing, is wrong and causes wrong epoll notification:
`s_UDTUnited.m_EPoll.update_events(m_SocketID, m_sPollID, UDT_EPOLL_OUT, true);`

- [ ] Fix `EPOLL_OUT` epoll update event.